llvm/lib/Transforms/Utils/SimplifyCFG.cpp

```diff
@@ -7104,6 +7104,47 @@ static bool simplifySwitchOfPowersOfTwo(SwitchInst *SI, IRBuilder<> &Builder,
   return true;
 }
 
+// Try to fold switch with manually selected default branch.
+// For example, for the switch
+//   switch (v) { case 0: case 1: case 2: ... case MaxC: default: }
+//     (continuous cases value from 0 to MaxC)
+//   and, v = select (x u<= MaxC), x, AnotherC.
+// so "AnotherC" is the de-facto default branch, and it will be folded to
+//   switch (x) { case 0: case 1: case 2: ... case MaxC:  default -> AnotherC: }
+static bool simplifySwitchWithImplicitDefault(SwitchInst *SI) {
+  Value *Cond;
+  ConstantInt *Bound, *DefaultCase;
+  ICmpInst::Predicate Pred;
+  if (!match(SI->getCondition(),
+             m_Select(m_ICmp(Pred, m_Value(Cond), m_ConstantInt(Bound)),
+                      m_Deferred(Cond), m_ConstantInt(DefaultCase))) ||
+      Pred != CmpInst::ICMP_ULT)
```

Comment:
Seems a bit constraining to only take `ICMP_ULT`, think at very least should handle `UGT` w/ the `DefaultCase` on the TrueArm.

Reply:
`icmp ugt + select + switch` doesn't exist. 
I think it is a rustc-specific pattern. https://github.com/rust-lang/rust/blob/81117ff930fbf3792b4f9504e3c6bccc87b10823/compiler/rustc_codegen_ssa/src/mir/place.rs#L206-L309

Reply:
Though it's possible that `icmp ult + select` will be transformed to `icmp ugt + select` by 2c75fe7f2a8b2, it seems not exist in practice. But I think it doesn't hurt to handle `ugt` just in case?

Reply:
Hmm, truthfully would just say make a helper that covers all unsigned preds. Shouldn't be that hard and then we don't need to spend what is probably more effort finding all the cases that each pred might be emitted in.